rclrs/src/time_source.rs

```diff
@@ -0,0 +1,167 @@
+use std::fmt;
+use std::sync::{Arc, Mutex};
+
+/// Time source for a node that subscribes to the /clock node, if the use_sim_time parameter is set
+pub struct TimeSource {
+    _node: Arc<Node>,
+    _clocks: Arc<Mutex<Vec<Arc<Mutex<Clock>>>>>,
+    _clock_qos: QoSProfile,
+    // TODO(luca) implement clock threads, for now will run in main thread
+    _use_clock_thread: bool,
+    // TODO(luca) Update this with parameter callbacks for use_sim_time
+    _ros_time_active: Arc<Mutex<bool>>,
+    _clock_subscription: Option<Arc<Subscription<ClockMsg>>>,
+    _last_time_msg: Arc<Mutex<Option<ClockMsg>>>,
+}
+
+pub struct TimeSourceBuilder {
+    node: Arc<Node>,
+    clock_qos: QoSProfile,
+    use_clock_thread: bool,
+    clock: Arc<Mutex<Clock>>,
+}
+
+impl TimeSourceBuilder {
+    pub fn new(node: Arc<Node>, clock: Arc<Mutex<Clock>>) -> Self {
+        Self {
+            node,
+            clock_qos: QOS_PROFILE_CLOCK,
+            use_clock_thread: true,
+            clock,
+        }
+    }
+
+    pub fn clock_qos(mut self, clock_qos: QoSProfile) -> Self {
+        self.clock_qos = clock_qos;
+        self
+    }
+
+    pub fn use_clock_thread(mut self, use_clock_thread: bool) -> Self {
+        self.use_clock_thread = use_clock_thread;
+        self
+    }
+
+    pub fn build(self) -> TimeSource {
+        let mut source = TimeSource {
+            _node: self.node.clone(),
+            _clocks: Arc::new(Mutex::new(vec![])),
+            _clock_qos: self.clock_qos,
+            _use_clock_thread: self.use_clock_thread,
+            _ros_time_active: Arc::new(Mutex::new(false)),
+            _clock_subscription: None,
+            _last_time_msg: Arc::new(Mutex::new(None)),
+        };
+        source.attach_clock(self.clock);
+        source.attach_node(self.node);
+        source
+    }
+}
+
+pub struct ClockMismatchError(ClockType);
+
+impl fmt::Display for ClockMismatchError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "Unexpected clock type {:?}, only RosTime can be used as a time source when use_sim_time is set",
+            self.0
+        )
+    }
+}
+
+impl TimeSource {
+    pub fn attach_clock(&self, clock: Arc<Mutex<Clock>>) -> Result<(), ClockMismatchError> {
+        let clock_type = clock.clone().lock().unwrap().clock_type();
+        if !matches!(clock_type, ClockType::RosTime) && *self._ros_time_active.lock().unwrap() {
+            return Err(ClockMismatchError(clock_type));
+        }
+        if let Some(last_msg) = self._last_time_msg.lock().unwrap().clone() {
+            let nanoseconds: i64 =
+                (last_msg.clock.sec as i64 * 1_000_000_000) + last_msg.clock.nanosec as i64;
+            Self::update_clock(&clock, nanoseconds);
+        }
+        // TODO(luca) this would allow duplicates to be stored in the vector but it seems other
+        // client libraries do the same, should we check and no-op if the value exists already?
```

Comment:
Perhaps an unordered set of some sort would work? That would make it so we don't have to check.

Reply:
Yea agreed, the hash set storage was one of my proposals in the [upstream issue](https://github.com/ros2/rclcpp/issues/2253), will check how that discussion goes and open matching PRs among the different client libraries to keep implementations consistent.

Reply:
Upstream consensus is also to move to a `HashSet` based approach. I tentatively did a vector based avoid duplication in https://github.com/ros2-rust/ros2_rust/pull/325/commits/1f5ea5cae786017d8b05b99a1741010a4253e9c8, however I think it would be tricky to have a `HashSet`  since `Arc` requires the inner type to be hashable and `rcl_clock_t` does not implement `Hash` or `PartialEq`. Hence we would have to write a custom Hash / Eq implementation, I would guess using `Arc::as_ptr` to hash based on the memory address.

Reply:
Simplified it in the end to get rid of the vector of clocks and just have a single clock.
To be honest even though the other client libraries allow having multiple clocks I haven't found a single case of that being done, so imho we can get away with the additional complexity of multiple clocks and just the node's clock.